Support loading a non-DP/DDP model from a DP/DDP state_dict

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53224 Support loading a non-DP/DDP model from a DP/DDP state_dict**

Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

Commit:
Support loading a non-DP/DDP model from a DP/DDP state_dict

Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

mrshenli commented on torch/nn/modules/module.py:
```diff
@@ -1181,7 +1181,10 @@ def load_state_dict(self, state_dict: 'OrderedDict[str, Tensor]',
             state_dict (dict): a dict containing parameters and
-                persistent buffers.
+                persistent buffers. To load a dict from a DataParallel or
+                DistributedDataParallel module, need to first apply
```
Is it possible to support this without changing `nn.Module` implementation? Otherwise, we are leaking a implementation detail of derived class into the base class. 

cc @gchanan @albanD

albanD replied:
Indeed, we either want to make this a public function or not document it. But not both.

SciPioneer replied:
The issue we want to address here is loading a DP/DDP model from a non-DP/DDP model, so this util function is prepared for non-DP/DDP model users.

albanD replied:
> so this util function is prepared for non-DP/DDP model users.

Well not really. Because to get into this situation, you have to be using a DP/DDP. Or actually know that you're loading a model based on DP/DDP. So this is only for users that are aware that they use these modules.

Other options I would consider:
- Make DP and DDP strip this "module" layer by default when saving. And add it back automatically when loading.
- Document this explicitly in the DP/DDP to tell users they should serialize the .module or use this utils function
- Prevent saving DP/DDP directly and redirect the users to save the .module instead

What do you think?

SciPioneer replied:
> Make DP and DDP strip this "module" layer by default when saving. And add it back automatically when loading.
I don't think this will work:
1) Some keys in DDP state_dict does not necessarily have the "module." prefix, so we cannot blindly add the prefix back for every key when loading.
2) If we change the loading logic in DDP, the DDP won't be compatible with any state_dict saved by previous DDP models.

> Document this explicitly in the DP/DDP to tell users they should serialize the .module or use this utils function
I think we can do this. You made a valid point that the user must be already aware that the saved model is a DDP model, so only adding the documentation in DDP should be fine.

> Prevent saving DP/DDP directly and redirect the users to save the .module instead
Again, this will hurt backward-compatibility if we prevent saving DP/DDP directly. We can add a soft suggestion in DDP documentation to courage users to save .module instead.

What do you think?

albanD replied:
> Some keys in DDP state_dict does not necessarily have the "module." prefix, so we cannot blindly add the prefix back for every key when loading.

Ho really? I thought that all the user modules are there and nothing else is around? 
And I don't think anyone in their right mind would subclass DP or DDP :p 

> If we change the loading logic in DDP, the DDP won't be compatible with any state_dict saved by previous DDP models.

We save versions in the serialization metadata just for that.
While loading, you would check the version, if it is old, load as-is, otherwise, implicitly append the `.module` before loading.

> I think we can do this. You made a valid point that the user must be already aware that the saved model is a DDP model, so only adding the documentation in DDP should be fine.

I think this is the right thing to do.
The only caveat is that there are so many warnings on these classes already, I am not sure users read them actually. Hence my more aggressive proposal to forbid this altogether. But this would be badly BC-breaking indeed. Note though that we could go through a deprecation cycle for this and then remove. @mrshenli do you think that is plausible? Or too many people rely on this?

SciPioneer replied:
See this line:
https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L526

The key in `state_dict._metadata` does not necessarily have a "module." prefix, so implicitly appending a prefix when loading can be tricky.

I moved the notes from `nn.Module` to DDP to address this documentation scope issue. Even if we want to take the more aggressive approach that forbids DDP from saving a state dict directly, I think we should do it in a separate PR.

albanD replied:
I agree that this should be a separate issue if we want to do it.

mrshenli commented on torch/nn/modules/module.py:
```diff
@@ -1181,7 +1181,10 @@ def load_state_dict(self, state_dict: 'OrderedDict[str, Tensor]',
-                persistent buffers.
+                persistent buffers. To load a dict from a DataParallel or
+                DistributedDataParallel module, need to first apply
+                :meth:`~torch.nn.modules.utils._consume_prefix_in_state_dict_if_present`
```
Is this intentional to expose a private function in a stable API doc?

albanD replied:
Indeed, we either want to make this a public function or not document it. But not both.

SciPioneer replied:
Made it public. Thanks!

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

rohan-varma commented on torch/nn/modules/utils.py:
```diff
@@ -32,3 +32,34 @@ def _list_with_default(out_size: List[int], defaults: List[int]) -> List[int]:
+            # 'module': for the actual model.
+            # 'module.xx.xx': for the rest.
+
+            if len(key) == 0:
```
The doc says we want to remove "" for the DDP module, which is when len(key) == 0, but here we just go back to the loop, seemingly without removing it?

SciPioneer replied:
If the metadata key is an empty string, there is no point to run `key[len(prefix) :]` remove the non-existing leading "module". I don't see any problem with this. We only need to remove "module" prefix if it exists in a key, right?

rohan-varma replied:
Agreed, but I'm still sort of confused by the comments above:
```
# for the metadata dict, the key can be:
            # '': for the DDP module, which we want to remove.
            # 'module': for the actual model.
            # 'module.xx.xx': for the rest.
```

if we want to remove "", why not remove it instead of `continue`?

rohan-varma commented on torch/nn/modules/utils.py:
```diff
@@ -32,3 +32,34 @@ def _list_with_default(out_size: List[int], defaults: List[int]) -> List[int]:
+
+            if len(key) == 0:
+                continue
+            newkey = key[len(prefix) :]
```
Do we not need to check if it startswith prefix here?

SciPioneer replied:
The comments here assumes that there are only 3 cases: as long as there is a prefix, the prefix must start with "module", then no need to check. This is my understanding.

I copied this code snippet from https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L503

I guess here the original author probably wants to save some overheads of unnecessary string comparison.

rohan-varma replied:
So every key in `metadata.keys()` is either `""` or starts with `module`? Are we sure that's always true for a DDP model, can we verify experimentally?

rohan-varma reviewed (commented):
It's good that we have a test in `test_nn`, but maybe let's also add one to `distributed_test` that directly calls `torch.save(model.state_dict)` on a DDP instance, and then verifies the model is loaded locally properly after a call to `consume_prefix_in_state_dict_if_present`?

rohan-varma commented on torch/nn/modules/utils.py:
```diff
@@ -32,3 +32,34 @@ def _list_with_default(out_size: List[int], defaults: List[int]) -> List[int]:
+
+    Args:
+        state_dict (OrderedDict): a state-dict to be loaded to the model.
+        prefix (str): prefix.
```
Maybe document how this is useful for ddp?

SciPioneer replied:
Added a note in the docstring.

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

SciPioneer commented:
> It's good that we have a test in `test_nn`, but maybe let's also add one to `distributed_test` that directly calls `torch.save(model.state_dict)` on a DDP instance, and then verifies the model is loaded locally properly after a call to `consume_prefix_in_state_dict_if_present`?

> So every key in metadata.keys() is either "" or starts with module? Are we sure that's always true for a DDP model, can we verify experimentally?

The logic of an either-""-or-"module." prefix is already used by a real use case under facebookresearch/fvcore repo. I also verified this logic by checking  the `state_dict` of a few DDP models.

Note that saving a DDP state_dict and loading it from a non-DDP model is also tested in `test_c10d.py` in this PR.

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

albanD commented on torch/nn/parallel/distributed.py:
```diff
@@ -229,6 +229,11 @@ class DistributedDataParallel(Module):
 
+    .. note::
+        To let a non-DDP model load a state dict from a DDP model,
+        need to first apply :meth:`~torch.nn.modules.utils.consume_prefix_in_state_dict_if_present`
```
nit: "you need" ?

albanD commented on torch/nn/modules/utils.py:
```diff
@@ -32,3 +32,39 @@ def _list_with_default(out_size: List[int], defaults: List[int]) -> List[int]:
+
+
+def consume_prefix_in_state_dict_if_present(state_dict: Dict[str, Any], prefix: str):
+    r"""Strip the prefix in state_dict metadata, if any.
```
nit: not sure if you need to have "metadata" here? you strip it in the whole state_dict, both data and metadata right?

SciPioneer replied:
Sounds good!

albanD commented on test/test_nn.py:
```diff
@@ -4365,15 +4365,24 @@ def test_load_state_dict(self):
+            self.assertEqual(len(incompatible_keys.missing_keys), 0)
+            self.assertEqual(len(incompatible_keys.unexpected_keys), 0)
+            self.assertNotIn('Incompatible', str(incompatible_keys))
+            self.assertNotIn('Incompatible', repr(incompatible_keys))
```
nit: this one is not needed as you check the string above.

albanD commented on test/test_nn.py:
```diff
@@ -4365,15 +4365,24 @@ def test_load_state_dict(self):
+            self.assertEqual(len(incompatible_keys.unexpected_keys), 0)
+            self.assertNotIn('Incompatible', str(incompatible_keys))
+            self.assertNotIn('Incompatible', repr(incompatible_keys))
+            self.assertEqual(net.linear1.weight.data, sd['linear1.weight'])
```
nit: no need to use `.data` here.
And in general, you should never use it :D

SciPioneer replied:
Thanks for catching this! It's from the legacy code in this file. It's a good time to update this.

albanD reviewed (approved):
LGTM
Thanks for doing the update!

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]

Commit:
Update on "Support loading a non-DP/DDP model from a DP/DDP state_dict"


Loading a DP/DDP dict just needs to strip the module prefix from all items in the state dict and the metadata.

One existing example is here: https://github.com/facebookresearch/fvcore/blob/master/fvcore/common/checkpoint.py#L239.

#Closes: https://github.com/pytorch/pytorch/issues/41048/

Differential Revision: [D26798495](https://our.internmc.facebook.com/intern/diff/D26798495/)

[ghstack-poisoned]
